src/Compilers/CSharp/Test/Emit2/Semantics/CollectionLiteralTests.cs

```diff
@@ -5644,6 +5667,2548 @@ private static void VerifyTypes(SemanticModel model, ExpressionSyntax expr, stri
+            Assert.False(originalType.HasCollectionBuilderAttribute(out _, out _));
+        }
+
+        [Fact]
+        public void CollectionBuilder_Retargeting()
+        {
+            string sourceA = """
+                using System;
+                using System.Collections;
+                using System.Collections.Generic;
+                using System.Runtime.CompilerServices;
+                [CollectionBuilder(typeof(MyCollectionBuilder), "Create")]
+                public struct MyCollection<T> : IEnumerable<T>
+                {
+                    public IEnumerator<T> GetEnumerator() => default;
+                    IEnumerator IEnumerable.GetEnumerator() => default;
+                }
+                public class MyCollectionBuilder
+                {
+                    public static void Create(int[] items) { }
+                }
+                """;
+            var comp = CreateCompilation(new[] { sourceA, s_constructibleCollectionAttributeDefinition }, targetFramework: TargetFramework.Mscorlib40);
+            var refA = comp.ToMetadataReference();
+
+            string sourceB = """
+                #pragma warning disable 219
+                class Program
+                {
+                    static void Main()
+                    {
+                        F([]);
+                        F(new());
+                    }
+                    static void F(MyCollection<int> c)
+                    {
+                    }
+                }
+                """;
+            comp = CreateCompilation(sourceB, references: new[] { refA }, targetFramework: TargetFramework.Mscorlib45);
+            comp.VerifyEmitDiagnostics(
+                // (6,11): error CS9181: Could not find an accessible 'Create' method with the expected signature.
+                //         F([]);
+                Diagnostic(ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound, "[]").WithArguments("Create").WithLocation(6, 11));
+
+            var collectionType = (NamedTypeSymbol)comp.GetMember<MethodSymbol>("Program.F").Parameters[0].Type;
+            Assert.Equal("MyCollection<System.Int32>", collectionType.ToTestDisplayString());
+            TypeSymbol builderType;
+            string methodName;
+            Assert.True(collectionType.HasCollectionBuilderAttribute(out builderType, out methodName));
+            Assert.Equal("MyCollectionBuilder", builderType.ToTestDisplayString());
+            Assert.Equal("Create", methodName);
+
+            var retargetingType = (RetargetingNamedTypeSymbol)collectionType.OriginalDefinition;
+            Assert.Equal("MyCollection<T>", retargetingType.ToTestDisplayString());
+            Assert.True(retargetingType.HasCollectionBuilderAttribute(out builderType, out methodName));
+            Assert.IsType<RetargetingNamedTypeSymbol>(builderType);
+            Assert.Equal("MyCollectionBuilder", builderType.ToTestDisplayString());
+            Assert.Equal("Create", methodName);
+        }
```

Comment:
Took a quick look through the tests but did not review in depth this round. I didn't see any circularity tests: please make sure we're testing `CollectionBuilderAttribute` using a collection expression as an argument applied to `CollectionBuilderAttribute`. I would also like to see tests for UnmanagedCallersOnly, CompilerFeatureRequired (you may need to create some IL for this one), and some tests with `await`s in element list.